### PR TITLE
GUI maintenance

### PR DESCRIPTION
**Pull request by davidgraeff**
* Dock widgets: Use scope settings and lambdas instead of private slots. Block signals for set methods.
* Rewrite MainWindow with a UI file.
* MainWindow: Only connect widget signals/slots. Use lambdas instead of private slots.
  Don't initialize hantekdsocontrol with user settings. This is done in main.cpp now.

Bigger picture:
* The core part should not rely on the graphical part to function -> For unit tests and implementation tests.
* All graphical interfaces should be migrated to UI files to separate interface code from controller logic.

### CODE REVIEW

File: openhantek/src/docks/SpectrumDock.cpp [hunk 1]
```diff
@@ -16,43 +17,56 @@
...
-    for (ChannelID channel = 0; channel < settings->scope.voltage.size(); ++channel) {
-        this->dockLayout->addWidget(this->usedCheckBox[channel], (int)channel, 0);
-        this->dockLayout->addWidget(this->magnitudeComboBox[channel], (int)channel, 1);
 
-        // Connect signals and slots
-        connect(usedCheckBox[channel], &QCheckBox::toggled, this, &SpectrumDock::usedSwitched);
-        QComboBox* cmb = magnitudeComboBox[channel];
-        connect(cmb, static_cast<void (QComboBox::*)(int)>(&QComboBox::currentIndexChanged), this, &SpectrumDock::magnitudeSelected);
+    // Initialize elements
+    for (ChannelID channel = 0; channel < scope->voltage.size(); ++channel) {
```
Comment (dendvz) [inline openhantek/src/docks/SpectrumDock.cpp]: Should be scope->spectrum.size() here. The majority of code below uses scope->spectrum, except usedCheckBox which should be labelled with channel name.

File: openhantek/src/mainwindow.ui [hunk 2]
```diff
@@ -0,0 +1,224 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (dendvz) [inline openhantek/src/mainwindow.ui]: I'm not so familiar with Qt restrictions; would be great to have entire .ui file stored in /res, not /src. There's no actual source code here, just layouts and action bindings.

File: openhantek/src/mainwindow.cpp [hunk 3]
```diff
@@ -47,250 +35,84 @@ OpenHantekMainWindow::OpenHantekMainWindow(HantekDsoControl *dsoControl, DataAna
...
     // Connect general signals
-    connect(this, &OpenHantekMainWindow::settingsChanged, this, &OpenHantekMainWindow::applySettings);
     connect(dsoControl, &HantekDsoControl::statusMessage, statusBar(), &QStatusBar::showMessage);
 
     // Connect signals to DSO controller and widget
-    connect(horizontalDock, &HorizontalDock::samplerateChanged, this, &OpenHantekMainWindow::samplerateSelected);
-    connect(horizontalDock, &HorizontalDock::timebaseChanged, this, &OpenHantekMainWindow::timebaseSelected);
+    connect(horizontalDock, &HorizontalDock::samplerateChanged, [this]() {
+        this->dsoControl->setSamplerate(this->settings->scope.horizontal.samplerate);
+        this->dsoWidget->updateSamplerate(this->settings->scope.horizontal.samplerate);
```
Comment (dendvz) [inline openhantek/src/mainwindow.cpp]: Do we really need all these detailed updateXXX() methods on DsoWidget side? We should probably emit MainWindow::settingsChanged() signal connected to applySettings() slots of DsoWidget and all dock windows.

Reason behind that is quite simple: DsoControl should be allowed to modify several attributes of settings object, hence multiple widgets might require update after.

File: openhantek/src/mainwindow.ui (hunk 2 above)
Comment (davidgraeff) [reply openhantek/src/mainwindow.ui]: The ui files contain xml and they get converted to cpp code. They express sources in a different way like in a web project where you have html, javascript, css in a src directory.

File: openhantek/src/mainwindow.cpp (hunk 3 above)
Comment (davidgraeff) [reply openhantek/src/mainwindow.cpp]: I think it's a misconception in the first place that the main window is responsible to keep DsoWidget and DsoControl in sync. And yes, the granularity is maybe to high. Lesser signals would be alright.

This PR is just about tidying up the Docks and converting the MainWindow though.

File: openhantek/src/docks/SpectrumDock.cpp (hunk 1 above)
Comment (davidgraeff) [reply openhantek/src/docks/SpectrumDock.cpp]: You are right. I will introduce a `scope->countChannels()` and harmonize the code in a follow-up PR.

**Pull request merged**